### PR TITLE
Fix poll_ready usage in ChainVerifier

### LINKED ISSUE #1679: Fix poll_ready usage for the ChainVerifier service
**Is your feature request related to a problem? Please describe.**

The `ChainVerifier` service misuses `poll_ready`. (See #1593.)

The code works right now. But if the `BlockVerifier` or `CheckpointVerifier` are changed, they could fill up `Buffer`s or `Batch`es, potentially causing hangs.

**Describe the solution you'd like**

Make sure each `poll_ready` is followed by a `call`:
- [x] Always return `Poll::Ready(Ok(()))` from `ChainVerifier::poll_ready`
    - you can delete the current implementation - it's wrong
- [x] Wrap `CheckpointVerifier` and `BlockVerifier` in `tower::Buffer`s
- [x] Use `ServiceExt::oneshot` to call `CheckpointVerifier` and `BlockVerifier`
- [x] Document the `poll_ready/call` invariants to avoid future changes causing hangs

The documentation should look something like:

```rust
// Correctness:
//
// We can't call `poll_ready` on the block and checkpoint verifiers here, because that leaks buffer slots.
// See #1593 for details.
```

See #1593 for a detailed background explanation.

**Describe alternatives you've considered**

Don't make any changes. Future changes might cause Zebra to hang.

**Follow Up**

Handle backpressure correctly (#1618)

**Original Issues**

Design: Make sure all poll_ready reservations are actually used in a call #1593 
Example: Fix poll_ready usage for the Inbound service #1620

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

Work in progress for https://github.com/ZcashFoundation/zebra/issues/1679

## Related Issues

Closes #1679.

Fixes part of #1675.

## Follow Up Work

Clippy will fail on this PR until #1706 merges.

### CODE REVIEW

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs]: This code calls `poll_ready` 3 times for each request:
* once for the `checkpoint` verifier in `poll_ready`
* once for the `block` verifier in `poll_ready`
* once for the chosen verifier in `call`

 `poll_ready` must only be called once per request, and every `poll_ready` must be followed by a `call`.

```suggestion
        // Correctness:
        //
        // We can't call `poll_ready` on the block and checkpoint verifiers here,
        // because each `poll_ready` must be followed by a `call`, and we don't
        // know which verifier we're going to choose yet.
        // See #1593 for details.
        Poll::Ready(Ok(()))
```

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs]: Please explain why you chose this bound for the buffer.

I think we might eventually need a bound of 4:
* sync service
* inbound service
* mining service (a feature we might add in future)
* 1 extra slot to avoid contention

We're going to use this bound for the block verifier and checkpoint verifier buffer. So let's turn it into a constant, and put an explanation in a doc comment on the constant.

Comment (teor2345) [reply zebra-consensus/src/chain.rs]: (Extra slots are pretty harmless, but missing slots can slow Zebra down a lot.)

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs:62]: Since we're using a `Buffer`, we get `BoxError`s back from the checkpoint and block verifiers.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs]: Oops
```suggestion
    S: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + Clone + 'static,
```

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs:80]: We can't call `poll_ready` here - it doesn't matter what we do with the results, it's the call that takes up a buffer slot.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs:102]: These services are now `Buffer`ed, so we don't need an `async` block. We can just call `ServiceExt::oneshot` on a cloned service, and return those futures.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs:159]: It's a bit easier to read the code if we create the services, then replace them with buffered versions.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [reply zebra-consensus/src/chain.rs:80]: @yaahc - this constraint is something to keep in mind for future service code reviews.

If we call `poll_ready`, but there isn't a matching `call`, we fill up our `Buffer`s with unused reservations, and Zebra hangs.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [reply zebra-consensus/src/chain.rs:102]: @yaahc @oxarbitrage the best way to avoid `poll_ready` / `call` mismatches is to just use a `oneshot`.

It handles the `poll_ready` and `call` internally, so we can't possibly get it wrong.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs:148]: This fix is part of #1675 - and it's actually something I missed during my review last week.

Since we're fixing this file, let's also use `ServiceExt::oneshot` here.

**Review by teor2345 — COMMENTED** [review]
I think this PR is fine now, but I wrote some of it, so @oxarbitrage or @yaahc should review it before it merges.

**Comment by oxarbitrage** [comment]
@teor2345 this looks great and thank you for all the comments, the logic looks good and the code very clean and concise.

I had a few problems with the references we had before to do this change.

1- https://github.com/ZcashFoundation/zebra/issues/1593 was too long and general. This is ok and all useful information but i got lost.
2- https://github.com/ZcashFoundation/zebra/pull/1620 makes more changes than are actually needed to implement `oneshot`. Got lost trying to understand them.
3- The issue https://github.com/ZcashFoundation/zebra/issues/1679 was suggesting initially a solution without `oneshot`. I was not smart enough to try to apply it anyways, it is a lot easier than  making the matches "manually".

However, with the content of this PR everything is a lot clear to me.

This PR teaches:

- how to buffer and use buffered services and errors.
- how to use `oneshot`.
- comments explain how to do `poll_ready` and match `call`s method.

**Comment by teor2345** [comment]
I'm re-running the jobs to make sure #1706 fixes clippy.

**Comment by teor2345** [comment]
I think clippy failed because it's working on the head commit, or an old merge.

Let's just merge, the type is used, so clippy should pass on `main`.

**Review by teor2345 — APPROVED** [review]
I wrote some of this, and others have checked it, so it's good.

**Pull request merged**
